pkg/scheduler/core/generic_scheduler.go

```diff
@@ -51,22 +51,71 @@ func (g *genericScheduler) Schedule(ctx context.Context, placement *policyv1alph
 		return result, fmt.Errorf("no clusters available to schedule")
 	}
 
-	feasibleClusters, err := g.findClustersThatFit(ctx, g.scheduleFramework, placement, &spec.Resource, clusterInfoSnapshot)
+	// 1. calculate reserved clusters and candidate clusters
+	// 2. prioritize candidate clusters
+	// 3. select target clusters from (reserved clusters + candidate clusters)
+	// 4. assign replicas
+	readyClusters, unreadyClusters, unjoinedClusters := clusterInfoSnapshot.InspectClusters()
+	bindClusters := util.ConvertToClusterNames(spec.Clusters)
+
+	reservedClusters := bindClusters.Intersection(readyClusters)
+	// Remove reserved clusters that don't fit the placement
+	for clusterName := range reservedClusters {
+		clusterObj := clusterInfoSnapshot.GetCluster(clusterName)
+		resMap := g.scheduleFramework.RunFilterPlugins(ctx, placement, &spec.Resource, clusterObj.Cluster())
+		res := resMap.Merge()
+		if !res.IsSuccess() {
+			klog.V(4).Infof("cluster %q is not fit", clusterName)
+			reservedClusters.Delete(clusterName)
+		}
+	}
+
+	availableClusters := readyClusters.Difference(bindClusters)
+	candidateClusters, err := g.findClustersThatFit(ctx, g.scheduleFramework, placement, &spec.Resource, clusterInfoSnapshot, availableClusters)
 	if err != nil {
 		return result, fmt.Errorf("failed to findClustersThatFit: %v", err)
 	}
-	if len(feasibleClusters) == 0 {
-		return result, fmt.Errorf("no clusters fit")
+
+	if len(candidateClusters) == 0 {
+		// just warn
+		klog.Warningf("There is no candidate clusters")
+	} else {
+		klog.V(4).Infof("Candidate clusters found: %v", candidateClusters)
 	}
-	klog.V(4).Infof("feasible clusters found: %v", feasibleClusters)
 
-	clustersScore, err := g.prioritizeClusters(ctx, g.scheduleFramework, placement, feasibleClusters)
+	unreadyTargetClusters := bindClusters.Intersection(unreadyClusters)
+	unjoinedTargetClusters := bindClusters.Intersection(unjoinedClusters)
```

Comment:
It seems like unjoinedTargetClusters should be bindClusters.Intersection(unjoinedClusters) + bindClusters.Difference(AllExistClusters)

Reply:
Thanks for correcting, fixed, sorry for my oversight😥

Reply:
> 

Take it easy, I'll finish my first-round review as soon as possible. The overall logic will be correct, but not elegant enough. We can keep improving it.